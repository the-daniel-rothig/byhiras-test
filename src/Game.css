:root {
  --red-text: #7d0c0c;
  --green-text: #4b6b06;
}

.character {
  position: absolute;
  bottom: 40%;
}

.character.left {
  left: 10%;
}

.character.right {
  right: 10%;
}

.character__sprite {
  display: flex;
  justify-content: center;
  padding-top: 30px;
  animation: hopping 800ms ease-in infinite;
}

.right .character__sprite {
  animation-delay: 200ms;
}

.character__sprite--dead {
  animation: dropOff 1s linear;
  transition-duration: 0.5s;
  transition-delay: 0.5s;
  transition-property: opacity;
  opacity: 0;
}

.lifebar {
  display: flex;
}

.lifebar__text {
  margin-left: 10px;
}

.lifebar__outer {
  width: 100px;
  height: 10px;
  border: solid black 1px;
  padding: 2px;
}

.lifebar__inner {
  background-color: red;
  width: 100px;
  height: 10px;
  transform-origin: 0%;
  transition: transform 200ms;
}


.swords, .dice {
  font-size: 5em;
  position: absolute;
  top: 30%;
  animation: pop 200ms linear;
}

.swords { 
  left: 50%;
  translate: -50%;
}

.dice {
  display: flex;
  flex-direction: column;
  line-height: 0.7;
}

.dice.left{
  left:40%;
  color: var(--green-text);
}

.dice.right { 
  right: 40%;
  color: var(--red-text);
}

.dice.dice--inline {
  display: inline-block;
  font-size: 2em;
  line-height: 0.5;
  position:relative;
  top: 10px;
  left: 0;
  right: 0;
  animation:none;
}

.die {
  animation: pop 200ms linear;
}

.attackButton, .gameMessage {
  position: absolute;
  bottom: 30%;
  left: 50%;
  translate: -50% 0;

  font-size: 1.5em;
  color: #523c02;
  border: solid 4px #9e7609;
  border-radius: 10px;
  background-color: #dbc179;
  padding: 0.5em 2em;
}

.attackButton {
  background-color:#b5953f;
  font-size: 2em;
}

.gameMessage {
  bottom:25%;
  animation: pop 200ms linear;
}

.gameMessage__good {
  font-size: 3em; 
  font-weight: bold;
  color: var(--green-text)
}

.gameMessage__bad {
  font-size: 3em; 
  font-weight: bold;
  color: var(--red-text);
}

@keyframes pop {
  0% { scale: 0 }
  50% { scale: 1.2}
  100% { scale: 1 }
}

@keyframes dropOff {
  0% { translate: 0 0; color: black}
  30% { translate: 0 -30px; color: red}
  100% { translate: 0 200px; }
}

@keyframes hopping {
  0% { translate: -5px 0 }
  35% { translate: 0 -5px }
  50% { translate: 5px 0 }
  85% { translate: 0 -5px }
  100% { translate: -5px 0 }
}