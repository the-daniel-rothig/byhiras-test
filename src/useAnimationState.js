import { useReducer, useRef, useEffect } from 'react';

function makeInitialState() {
  return {
    active: false,
    showSwords: false,
    showDice: false,
    showDamage: false
  }
}

const ACTION_TYPES = { 
  ADVANCE: "ADVANCE",
  RESET: "RESET"
}

function reduce(state, action) {
  switch (action.type) {
    case ACTION_TYPES.ADVANCE: 
    return {
      active: !state.showDamage,
      showSwords: !state.showDamage && state.active,
      showDice: !state.showDamage && state.showSwords,
      showDamage: !state.showDamage && state.showDice,
    }
    case ACTION_TYPES.RESET:
      return makeInitialState();
    default:
      throw `unknown action type: ${action.type}`
  }
}

export default () => {
  const timeouts = useRef([]);
  const [ state, dispatch ] = useReducer(reduce, makeInitialState());

  function clearTimeouts() {
    timeouts.current.forEach(to => clearTimeout(to));
    timeouts.current = [];
  }

  useEffect(() => () => clearTimeouts(), []);

  function stop() {
    clearTimeouts();
    dispatch({type: ACTION_TYPES.RESET});
  }

  function start() {
    stop();
    dispatch({type: ACTION_TYPES.ADVANCE});
    timeouts.current = [
      setTimeout(() => dispatch({type: ACTION_TYPES.ADVANCE}), 100),
      setTimeout(() => dispatch({type: ACTION_TYPES.ADVANCE}), 300),
      setTimeout(() => dispatch({type: ACTION_TYPES.ADVANCE}), 800),
      setTimeout(() => dispatch({type: ACTION_TYPES.ADVANCE}), 2500)
    ];
  }

  return { state, start, stop };
}