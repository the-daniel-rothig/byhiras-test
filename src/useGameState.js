import { useReducer } from 'react';

const ACTION_TYPES = {
  ATTACK: "ATTACK"
};

function reduce(state, action) {
  switch (action.type) {
    case ACTION_TYPES.ATTACK:
      const heroDice = [ throwDie(), throwDie() ];
      const monsterDice = [ throwDie(), throwDie() ];

      const score = sum(heroDice) - sum(monsterDice);
      return {
        oldHeroHp: state.heroHp,
        heroHp: Math.max(state.heroHp + Math.min(score, 0), 0),
        oldMonsterHp: state.monsterHp,
        monsterHp: Math.max(state.monsterHp - Math.max(0, score), 0),
        heroDice,
        monsterDice
      }
      
    default:
      throw `unknown action type ${action.type}.`;
  }
}

function sum(array) {
  return array.reduce((a,b) => a+b, 0);
}

function throwDie() {
  return Math.ceil(6 * Math.random());
}

export const INITIAL_HP = 100;

function makeInitialState() {
  return {
    oldHeroHp: INITIAL_HP,
    heroHp: INITIAL_HP,
    oldMonsterHp: INITIAL_HP,
    monsterHp: INITIAL_HP,
    heroDice: [],
    monsterDice: [],
  }
}

export default () => {
  const [state, dispatch] = useReducer(reduce, makeInitialState());
  
  return {
    attack: () => dispatch({type: ACTION_TYPES.ATTACK}),
    state
  };
}