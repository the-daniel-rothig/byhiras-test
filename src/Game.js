import React from 'react';
import useGameState, { INITIAL_HP } from './useGameState';
import useAnimationState from './useAnimationState';
import './Game.css';
import cn from 'classnames';

const Character = ({className, hp, children}) => {
  const lifeScale = Math.max(hp / INITIAL_HP, 0); 

  return (
    <div className={cn("character", className)}>
      <div className="lifebar">
        <div className={"lifebar__outer"}>
          <div className={"lifebar__inner"} style={{transform: `scaleX(${lifeScale})`}} />
        </div>
        <div class="lifebar__text">{Math.max(0, hp)} HP</div> 
      </div>
      <div className={cn("character__sprite", {"character__sprite--dead": hp <= 0})}>
        <div>
          {children}
        </div>
      </div>
    </div>
  )
}

const Hero = () => (
  <pre role="img" aria-label="our hero">{`
|_O       
  |\`-)--- 
  |\\      
/  |
  `}</pre>
);

const Monster = () => (
  <pre role="img" aria-label="a monster">{`
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 
`}</pre>
);

const Swords = () => <span role="img" aria-label="swords clash!" className={"swords"}>⚔️</span>

const dieFaces = [
  "⚀",
  "⚁",
  "⚂",
  "⚃",
  "⚄",
  "⚅"
];

const Dice = ({className, values}) => (
  <div className={cn("dice", className)}>
    {values.map(v => (
      <span role="img" aria-label={`die of value ${v}`} className={"die"}>
        {dieFaces[v - 1]}
      </span>
    ))}
  </div>
)

const DamageMessage = ({gameState}) => {
  let message = "You parried each other's blows.";

  if (gameState.oldHeroHp !== gameState.heroHp) {
    const damage = gameState.oldHeroHp - gameState.heroHp;

    message = (
      <>
        {damage > 5 ? "OUCH! " : ""}
        You've taken{" "}
        <Dice className="dice--inline right" values={gameState.monsterDice} />
        {" "}-{" "}
        <Dice className="dice--inline left" values={gameState.heroDice} />
        {" "}= {damage} damage.
      </>
    );
  }

  if (gameState.oldMonsterHp !== gameState.monsterHp) {
    const damage = gameState.oldMonsterHp - gameState.monsterHp
    message = (
      <>
        {damage > 5 ? "WOW! " : ""}
        You've dealt{" "}
        <Dice className="dice--inline left" values={gameState.heroDice} />
        {" "}-{" "}
        <Dice className="dice--inline right" values={gameState.monsterDice} />
        {" "}= {damage} damage.
      </>
    );
  }

  return (
    <span className={"gameMessage"}>
      {message}
    </span>
  );
}

const GameOver = ({win}) => (
  <span className={cn("gameMessage", {"gameMessage__good": win, "gameMessage__bad": !win})}>
    {win ? "You Win" : "Game Over"}
  </span>
)

const Game = () => {
  const game = useGameState();
  const animation = useAnimationState();

  const showOldHp = animation.state.active && !animation.state.showDamage;

  const heroHp = showOldHp ? game.state.oldHeroHp : game.state.heroHp;
  const monsterHp = showOldHp ? game.state.oldMonsterHp : game.state.monsterHp;

  const gameOver = heroHp <= 0 || monsterHp <= 0;

  function onClick() {
    game.attack();
    animation.start();
  }

  return (
    <>
      <Character hp={heroHp} className={"left"}>
        <Hero />
      </Character>
      <Character hp={monsterHp} className={"right"}>
        <Monster />
      </Character>
      {animation.state.showSwords && <Swords />}
      {animation.state.showDice && <Dice className={"left"} values={game.state.heroDice} />}
      {animation.state.showDice && <Dice className={"right"} values={game.state.monsterDice} />}
      {animation.state.showDamage && <DamageMessage gameState={game.state}/>}

      {!animation.state.active && !gameOver && <button className={"attackButton"} onClick={onClick}>Attack!</button>}
      {!animation.state.active && gameOver && <GameOver win={heroHp > 0} />}
    </>
  )
}

export default Game;