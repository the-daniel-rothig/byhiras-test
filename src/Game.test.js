import React from 'react';
import { render } from '@testing-library/react';
import './setupTests';

import useAnimationState from './useAnimationState';
import useGameState from './useGameState';

import Game from './Game';

jest.mock('./useAnimationState', () => {
  let state = {};
  const startMock = jest.fn();
  const fn = jest.fn(() => ({
    state,
    start: startMock
  }))
  fn.set = newState => state = newState;
  fn.startMock = startMock;

  return fn;
});

jest.mock('./useGameState', () => {
  let state = {};
  const attackMock = jest.fn();
  const fn = jest.fn(() => ({
    state,
    attack: attackMock
  }))
  fn.set = newState => state = newState;
  fn.attackMock = attackMock;

  return fn;
})


describe('Game', () => {
  const defaultAnimationState = {
    active: false,
    showSwords: false,
    showDice: false,
    showDamage: false
  };

  const defaultGameState = {
    oldHeroHp: 100,
    heroHp: 100,
    oldMonsterHp: 51,
    monsterHp: 50,
    heroDice: [1,2],
    monsterDice: [1,1]
  }

  beforeEach(() => {
    useAnimationState.set(defaultAnimationState);
    useGameState.set(defaultGameState);
  })

  it('renders the initial state', () => {
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('launches attack when the attack button is pressed', () => {
    const { getByText } = render(<Game />);
    getByText("Attack!").click();
    expect(useAnimationState.startMock).toHaveBeenCalled();
    expect(useGameState.attackMock).toHaveBeenCalled();
  });

  it('renders correctly on animation start', () => {
    useAnimationState.set({
      ...defaultAnimationState, 
      active: true
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  })

  it('renders correctly when the swords show', () => {
    useAnimationState.set({
      ...defaultAnimationState, 
      active: true,
      showSwords: true,
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly when the dice show', () => {
    useAnimationState.set({
      ...defaultAnimationState, 
      active: true,
      showSwords: true,
      showDice: true
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly when the damage shows', () => {
    useAnimationState.set({
      ...defaultAnimationState, 
      active: true,
      showSwords: true,
      showDice: true,
      showDamage: true
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('renders a Game Over correctly', () => {
    useGameState.set({
      ...defaultGameState,
      heroHp: 0
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });

  it('renders a You Won correctly', () => {
    useGameState.set({
      ...defaultGameState,
      monsterHp: 0
    });
    const { container } = render(<Game />);
    expect(container).toMatchSnapshot();
  });


})