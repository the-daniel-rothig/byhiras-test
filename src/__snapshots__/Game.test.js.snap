// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game renders a Game Over correctly 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        0
         HP
      </div>
    </div>
    <div
      class="character__sprite character__sprite--dead"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        50
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <span
    class="gameMessage gameMessage__bad"
  >
    Game Over
  </span>
</div>
`;

exports[`Game renders a You Won correctly 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        0
         HP
      </div>
    </div>
    <div
      class="character__sprite character__sprite--dead"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <span
    class="gameMessage gameMessage__good"
  >
    You Win
  </span>
</div>
`;

exports[`Game renders correctly on animation start 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        51
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
</div>
`;

exports[`Game renders correctly when the damage shows 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        50
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <span
    aria-label="swords clash!"
    class="swords"
    role="img"
  >
    ⚔️
  </span>
  <div
    class="dice left"
  >
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
    <span
      aria-label="die of value 2"
      class="die"
      role="img"
    >
      ⚁
    </span>
  </div>
  <div
    class="dice right"
  >
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
  </div>
  <span
    class="gameMessage"
  >
    
    You've dealt
     
    <div
      class="dice dice--inline left"
    >
      <span
        aria-label="die of value 1"
        class="die"
        role="img"
      >
        ⚀
      </span>
      <span
        aria-label="die of value 2"
        class="die"
        role="img"
      >
        ⚁
      </span>
    </div>
     
    -
     
    <div
      class="dice dice--inline right"
    >
      <span
        aria-label="die of value 1"
        class="die"
        role="img"
      >
        ⚀
      </span>
      <span
        aria-label="die of value 1"
        class="die"
        role="img"
      >
        ⚀
      </span>
    </div>
     
    = 
    1
     damage.
  </span>
</div>
`;

exports[`Game renders correctly when the dice show 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        51
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <span
    aria-label="swords clash!"
    class="swords"
    role="img"
  >
    ⚔️
  </span>
  <div
    class="dice left"
  >
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
    <span
      aria-label="die of value 2"
      class="die"
      role="img"
    >
      ⚁
    </span>
  </div>
  <div
    class="dice right"
  >
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
    <span
      aria-label="die of value 1"
      class="die"
      role="img"
    >
      ⚀
    </span>
  </div>
</div>
`;

exports[`Game renders correctly when the swords show 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        51
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <span
    aria-label="swords clash!"
    class="swords"
    role="img"
  >
    ⚔️
  </span>
</div>
`;

exports[`Game renders the initial state 1`] = `
<div>
  <div
    class="character left"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        100
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="our hero"
          role="img"
        >
          
|_O       
  |\`-)--- 
  |\\      
/  |
  
        </pre>
      </div>
    </div>
  </div>
  <div
    class="character right"
  >
    <div
      class="lifebar"
    >
      <div
        class="lifebar__outer"
      >
        <div
          class="lifebar__inner"
          style="transform: scaleX(NaN);"
        />
      </div>
      <div
        class="lifebar__text"
      >
        50
         HP
      </div>
    </div>
    <div
      class="character__sprite"
    >
      <div>
        <pre
          aria-label="a monster"
          role="img"
        >
          
oo\`'._..---.___..-
(_,-.        ,..'\`
    \`'.    ;
        : :\`
      _;_; 

        </pre>
      </div>
    </div>
  </div>
  <button
    class="attackButton"
  >
    Attack!
  </button>
</div>
`;
